src/compiler/emitter.ts

```diff
@@ -5372,7 +5349,30 @@ export function createPrinter(printerOptions: PrinterOptions = {}, handlers: Pri
     }
 
     function isReservedName(name: string, privateName: boolean): boolean {
-        return privateName ? !!reservedPrivateNames?.has(name) : !!reservedNames?.has(name);
+        let set: Set<string> | undefined;
+        let stack: (Set<string> | undefined)[];
+        if (privateName) {
+            set = reservedPrivateNames;
+            stack = reservedPrivateNamesStack;
+        }
+        else {
+            set = reservedNames;
+            stack = reservedNamesStack;
+        }
+
+        if (set?.has(name)) {
+            return true;
+        }
+        for (let i = stack.length - 1; i >= 0; i--) {
+            if (set === stack[i]) {
+                continue;
+            }

```

Comment:
This is partially off-topic, but is there any reason the current scope can't be stored separate of the stack itself? Or if it is possible, when is it the case that the current is not the same as the last in the stack?

Reply:
The current scope is stored separate from the stack. The way push/pop works is that we don't reset the current scope until we actually record something so that we defer the cost of reassignment until the moment we actually need to record a name. This check here is an example of that. If the set for the current scope is also in the stack, then we can skip it as nothing else was actually recorded in that scope.

Reply:
In other words, the algorithm favors code that requires no generated names so that we can avoid paying the cost for more recent `--target`s.